# Exploratory Data Analysis 

## Objectives

Students should be able to:

- Identify key questions when investigating a new data set.
- Answer the questions identified by computing appropriate metrics and creating data visualizations.
- Apply appropriate data visualization techniques (box plots, histograms, scatter, bar charts, heat maps, etc.) to answer key questions about a data set.
- Calculate appropriate metrics of features/and or targets to answer questions (mean, number of missing values per feature, etc.)


### Question:

Now lets put these skills together in the way we would typically explore data. This is called EDA. Exploratory Data Analysis.

Open chess_games.csv and answer these questions (use markdown headers to denote which question you are answering) 

1. how many rows of data do you have?
1. What does each row represent?
1. Who won more, white or black?
1. How many moves per game on average?
1. What was the most likely 'first move'?
1. How many games end in checkmate? (the alternative is to surrender or timeout)
1. What percent of games is that?
1. How long was the average game?
1. How long was the average game that white won? How long was the average game that was a draw?
1. What is an increment code?
1. Ask three of your own questions *in writing*. Then answer them. (we write our questions because if you come back later you'll forget what question you were answering)


write your question after the '####':

#### 1: [YOUR QUESTION TITLE HERE]

In [11]:
#  how many rows of data do you have?
# write code here:
chess = pd.read_csv('data/chess_games.csv')
chess.count()


id                20058
rated             20058
created_at        20058
last_move_at      20058
turns             20058
victory_status    20058
winner            20058
increment_code    20058
white_id          20058
white_rating      20058
black_id          20058
black_rating      20058
moves             20058
opening_eco       20058
opening_name      20058
opening_ply       20058
dtype: int64

1. How many rows of data do you have? 20058

<details><summary>Solution 1
</summary>

```python
chess = pd.read_csv('data/chess_games.csv')
chess.count()
```

</details>

write your question after the '####':

#### 

In [9]:
# What does each row represent?
# write code here:
chess['moves']
chess['winner']
chess['turns']
chess['victory_status']
chess.T


0      \
id                                                       TZJHLljE   
rated                                                       False   
created_at                                        1504210000000.0   
last_move_at                                      1504210000000.0   
turns                                                          13   
victory_status                                          outoftime   
winner                                                      white   
increment_code                                               15+2   
white_id                                                 bourgris   
white_rating                                                 1500   
black_id                                                     a-00   
black_rating                                                 1191   
moves           d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...   
opening_eco                                                   D10   
opening_name                     Slav Defense: Exchange Variation   
opening_ply                                                     5   

                                                            1      \
id                                                       l1NXvwaE   
rated                                                        True   
created_at                                        1504130000000.0   
last_move_at                                      1504130000000.0   
turns                                                          16   
victory_status                                             resign   
winner                                                      black   
increment_code                                               5+10   
white_id                                                     a-00   
white_rating                                                 1322   
black_id                                                skinnerua   
black_rating                                                 1261   
moves           d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...   
opening_eco                                                   B00   
opening_name               Nimzowitsch Defense: Kennedy Variation   
opening_ply                                                     4   

                                                            2      \
id                                                       mIICvQHh   
rated                                                        True   
created_at                                        1504130000000.0   
last_move_at                                      1504130000000.0   
turns                                                          61   
victory_status                                               mate   
winner                                                      white   
increment_code                                               5+10   
white_id                                                   ischia   
white_rating                                                 1496   
black_id                                                     a-00   
black_rating                                                 1500   
moves           e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...   
opening_eco                                                   C20   
opening_name                King's Pawn Game: Leonardis Variation   
opening_ply                                                     3   

                                                            3      \
id                                                       kWKvrqYL   
rated                                                        True   
created_at                                        1504110000000.0   
last_move_at                                      1504110000000.0   
turns                                                          61   
victory_status                                               mate   
winner                                                 

2. What does each row represent? The rows show information chess match including games, players, winners of each match

<details><summary>Solution 2
</summary>

```chess['moves']``` shows that we have multiple moves per row. So probably games? Ah, and theres a winner in ```chess['winner']```. Games.
    
HINT: ```chess.T``` switches rows and columns. (it stands for transpose). Using it allows us to see more of the columns in one look.
   
```python

```

</details>

write your question after the '####':

#### 

In [16]:
# Who won more, white or black?
# write code here:
chess['winner'].value_counts('white')


<bound method IndexOpsMixin.value_counts of 0        white
1        black
2        white
3        white
4        white
         ...  
20053    white
20054    black
20055    white
20056    white
20057    black
Name: winner, Length: 20058, dtype: object>

3. Who won more, white or black? white

<details><summary>Solution 3
</summary>

   
```python
chess['winner'].value_counts()

```

</details>

write your question after the '####':

####

In [17]:
# How many moves per game on average?
# write code here:
chess.T
chess['moves']
chess['winner']
chess.moves.apply(lambda x: x.split(' '))
chess.moves.apply(lambda x: len(x.split(' ')))
chess.moves.apply(lambda x: len(x.split(' '))).mean()

60.46599860404826

4) How many moves per game on average? 60

<details><summary>Solution 4
</summary>

```chess.T``` allows us to see more of the rows.

```chess['moves']``` shows that we have multiple moves per row. So probably games? Ah, and theres a winner in ```chess['winner']```. Games.
    
```python

>>> chess.moves.apply(lambda x: x.split(' ')) # shows the items per row in a list

0        [d4, d5, c4, c6, cxd5, e6, dxe6, fxe6, Nf3, Bb...
1        [d4, Nc6, e4, e5, f4, f6, dxe5, fxe5, fxe5, Nx...
2        [e4, e5, d3, d6, Be3, c6, Be2, b5, Nd2, a5, a4...
3        [d4, d5, Nf3, Bf5, Nc3, Nf6, Bf4, Ng4, e3, Nc6...
4        [e4, e5, Nf3, d6, d4, Nc6, d5, Nb4, a3, Na6, N...
                               ...                        

>>> chess.moves.apply(lambda x: len(x.split(' '))) # shows the length per list

0         13
1         16
2         61
3         61
4         95

>>> chess.moves.apply(lambda x: len(x.split(' '))).mean() # take the mean
60.46599860404826
```

</details>

write your question after the '####':

#### 

In [18]:
# What is the most common opening move?
# write code here:
chess.moves.apply(lambda x: x.split(' ')[0]).value_counts()


e4     12598
d4      4522
Nf3      725
c4       716
e3       416
g3       186
b3       173
f4       166
d3       131
Nc3       99
b4        88
c3        56
g4        38
h4        33
a4        28
a3        27
f3        23
Nh3       15
h3        14
Na3        4
Name: moves, dtype: int64

5. What was the most likely 'first move'? e4

<details><summary>Solution 5
</summary>

```python
chess.moves.apply(lambda x: x.split(' ')[0]).value_counts()
```

</details>

write your question after the '####':

#### 

In [19]:
# How many games end in checkmate?
# write code here:
chess.moves.apply(lambda x: x.split(' ')[-1]).value_counts()
chess.moves.apply(lambda x: x.split(' ')[-1][-1]).value_counts()

#    6325
+    2764
5    1890
4    1806
6    1599
3    1466
7    1185
2    1101
1     879
8     844
Q     130
O      62
R       4
N       2
B       1
Name: moves, dtype: int64

6. How many games end in checkmate? (the alternative is to surrender or timeout) 6325

<details><summary>Solution 6
</summary>


```python
>>> chess.moves.apply(lambda x: x.split(' ')[-1]).value_counts()

Qxf7#    176
Qg7#     170
Qg2#     157
Qxg2#    141
Qxg7#    137
        ... 
Qhh4#      1
Bxg6#      1
Rxa2+      1
Qac8#      1
Ngxe3      1

# hmm. Now I had to go look up what '#' means. Ok it means Checkmate. So now I grab the end of each string.

>>> chess.moves.apply(lambda x: x.split(' ')[-1][-1]).value_counts()

#    6325
+    2764
5    1890
4    1806
6    1599
3    1466
7    1185
2    1101
1     879
8     844
Q     130
O      62
R       4
N       2
B       1

```

</details>

In [20]:
# What percent of games end in checkmate?
# write code here:
chess.moves.apply(lambda x: x.split(' ')[-1][-1]).value_counts()/len(chess)


#    0.315336
+    0.137800
5    0.094227
4    0.090039
6    0.079719
3    0.073088
7    0.059079
2    0.054891
1    0.043823
8    0.042078
Q    0.006481
O    0.003091
R    0.000199
N    0.000100
B    0.000050
Name: moves, dtype: float64

7. What percent of games is that? ~32% (0.315336)

<details><summary>Solution 7
</summary>


```python
>>> chess.moves.apply(lambda x: x.split(' ')[-1][-1]).value_counts()/len(chess)

#    0.315336
+    0.137800
5    0.094227
4    0.090039
6    0.079719
3    0.073088
7    0.059079
2    0.054891
1    0.043823
8    0.042078
Q    0.006481
O    0.003091
R    0.000199
N    0.000100
B    0.000050
Name: moves, dtype: float64
```
> 31% of games ended with checkmate, and 13% of games ended with a normal 'check'
</details>

In [21]:
# How long was the average game?
# write code here:
chess['datetime_last_move'] = pd.to_datetime(chess['last_move_at'], unit='ms')
chess['datetime_created_at'] = pd.to_datetime(chess['created_at'], unit='ms')
chess['datetime_last_move'] - chess['datetime_created_at']


0             0.0
1             0.0
2             0.0
3             0.0
4             0.0
           ...   
20053    321734.0
20054    744219.0
20055    172834.0
20056    946699.0
20057    736699.0
Length: 20058, dtype: float64

In [25]:
chess['datetime_last_move'] = pd.to_datetime(chess['last_move_at'], unit='ms')
chess['datetime_created_at'] = pd.to_datetime(chess['created_at'], unit='ms')
(chess['datetime_last_move'] - chess['datetime_created_at']).mean()

Timedelta('0 days 00:14:29.707049606')

In [ ]:
8. How long was the average game? 14 minutes, 30 seconds

<details><summary>Solution 8
</summary>

```python
>>> chess['last_move_at'] - chess['created_at']

0             0.0
1             0.0
2             0.0
3             0.0
4             0.0
           ...   
20053    321734.0
20054    744219.0
20055    172834.0
20056    946699.0
20057    736699.0
Length: 20058, dtype: float64

# hmmm. This gives a bunch of... nanoseconds? miliseconds?"

# I netter google 'how to go from milliseconds to pandas datetime.'
# I click stack overflow, The first suggestion (and always a good bet).
# First answer says pd.to_datetime(df['UNIXTIME'], unit='ms')
# Lets try that.


>>> pd.to_datetime(chess['datetime_created_at'], unit='ms')

0       2017-08-31 20:06:40.000000000
1       2017-08-30 21:53:20.000000000
2       2017-08-30 21:53:20.000000000
3       2017-08-30 16:20:00.000000000
4       2017-08-29 18:06:40.000000000
                     ...             
20053   2017-07-11 16:35:14.342000128
20054   2017-07-10 14:48:09.760000000
20055   2017-07-10 14:44:37.492999936
20056   2017-07-10 14:15:27.019000064
20057   2017-07-09 23:32:32.648999936
Name: datetime_created_at, Length: 20058, dtype: datetime64[ns]

That looks reasonable

>>> chess['datetime_last_move'] = pd.to_datetime(chess['last_move_at'], unit='ms')
>>> chess['datetime_created_at'] = pd.to_datetime(chess['created_at'], unit='ms')
>>> chess['datetime_last_move'] - chess['datetime_created_at']

0              00:00:00
1              00:00:00
2              00:00:00
3              00:00:00
4              00:00:00
              ...      
20053   00:05:21.734000
20054   00:12:24.219000
20055   00:02:52.834000
20056   00:15:46.699000
20057   00:12:16.699000
Length: 20058, dtype: timedelta64[ns]


>>> (chess['datetime_last_move'] - chess['datetime_created_at']).mean()

Timedelta('0 days 00:14:29.707049')

# looks like the average game is '14 minutes, 30 seconds'

```

</details>

In [26]:
# How long was the average game that white won? How long was the average game that was a draw?
# write code here:
chess['game_ms'] = chess['last_move_at'] - chess['created_at']
pd.to_datetime(chess.groupby('winner').mean()['game_ms'], unit='ms')


winner
black   1970-01-01 00:13:54.751459317
draw    1970-01-01 00:23:18.628142105
white   1970-01-01 00:14:11.295443055
Name: game_ms, dtype: datetime64[ns]

9. How long was the average game that white won? 14 minutes, 11 seconds
How long was the average game that was a draw? 23 minutes, 18 seconds

<details><summary>Solution 9
</summary>


```python
# this one was hard for me because
chess.groupby('winner').mean()
# Doesn't return our column! Bleh!

# but this works
chess[chess['winner']=='white'].mean()

# and so does this
chess['game_ms'] = chess['last_move_at'] - chess['created_at']
pd.to_datetime(chess.groupby('winner').mean()['game_ms'], unit='ms')
```
    
</details>

In [ ]:
# write code here:



In [ ]:
# What is the increment code? Do not not know what this question is referring?

<details><summary> Solution 10
</summary>

google 'increment code chess':

"increment (in seconds) is the amount added after each move."

</details>

Now do your own! Make them headers, in markdown, in writing!

How many chess matches did skinnerua compete?

In [33]:
chess['black_id'].value_counts('skinnerua')
chess['white_id'].value_counts('skinnerua')


taranga          0.003590
chess-brahs      0.002642
a_p_t_e_m_u_u    0.002443
bleda            0.002393
ssf7             0.002393
                   ...   
zzzimon          0.000050
chessplay77      0.000050
mkmarijan        0.000050
nicomoli         0.000050
tomazkrampf      0.000050
Name: white_id, Length: 9438, dtype: float64

In [33]:
chess['black_id'].value_counts('skinnerua')
chess['white_id'].value_counts('skinnerua')


taranga          0.003590
chess-brahs      0.002642
a_p_t_e_m_u_u    0.002443
bleda            0.002393
ssf7             0.002393
                   ...   
zzzimon          0.000050
chessplay77      0.000050
mkmarijan        0.000050
nicomoli         0.000050
tomazkrampf      0.000050
Name: white_id, Length: 9438, dtype: float64

In [37]:
skinnerua_black = chess['black_id'].value_counts('skinnerua')
skinnerua_white = chess['white_id'].value_counts('skinnerua')
print(skinnerua_black)
print(skinnerua_white)

taranga               0.004088
vladimir-kramnik-1    0.002991
a_p_t_e_m_u_u         0.002343
king5891              0.002194
docboss               0.002194
                        ...   
stob1                 0.000050
eitaho                0.000050
i_see_u               0.000050
arrchess              0.000050
gideoni               0.000050
Name: black_id, Length: 9331, dtype: float64
taranga          0.003590
chess-brahs      0.002642
a_p_t_e_m_u_u    0.002443
bleda            0.002393
ssf7             0.002393
                   ...   
zzzimon          0.000050
chessplay77      0.000050
mkmarijan        0.000050
nicomoli         0.000050
tomazkrampf      0.000050
Name: white_id, Length: 9438, dtype: float64


In [39]:
9331 + 9438

18769

In [ ]:
How many chess matches did skinnerua compete? 18769 chess matches

How many chess matches open with Slav Defense: Exchange Variation?

In [34]:
chess['opening_name'].value_counts('Slav Defense: Exchange Variation')

Van't Kruijs Opening                                                 0.018347
Sicilian Defense                                                     0.017848
Sicilian Defense: Bowdler Attack                                     0.014757
French Defense: Knight Variation                                     0.013511
Scotch Game                                                          0.013511
                                                                       ...   
Modern Defense: Lizard Defense |  Mittenberger Gambit                0.000050
Queen's Gambit Declined: Alapin Variation                            0.000050
Queen's Indian Defense: Classical Variation |  Polugaevsky Gambit    0.000050
Polish Opening: Grigorian Variation                                  0.000050
Benko Gambit Accepted |  Yugoslav |  without 7...Bxf1                0.000050
Name: opening_name, Length: 1477, dtype: float64

How many chess matches open with Slav Defense: Exchange Variation? 1477 chess matches

What is the difference in turns between the shortest and longest chess matches?

### Question:

Now let's put these skills together AGAIN in the way we would typically explore data. This is called EDA. Exploratory Data Analysis.

Open exo.csv, skim https://exoplanetarchive.ipac.caltech.edu/docs/API_kepcandidate_columns.html and answer these questions.


This time write your own markdown headers before you answer each question!

1. how many rows of data do you have?
1. What does each row represent?
1. How many planets are confirmed?
1. What is the maximum temperature of the planets?
1. What is the maximum temperature of the stars?
1. What is the distance between the planet and the star?
1. How many planets are in each system?
1. How much does the planet temperatures vary?
1. How much does the star temperatures vary?
1. How many NaNs are in each category?
1. Ask three of your own questions *in writing*. Then answer them. (we write our questions because if you come back later you'll foget what question you were answering)


No solutions provided. You got this.

In [43]:
exo = pd.read_csv('data/exo.csv')
exo.count()

FileNotFoundError: [Errno 2] No such file or directory: 'data/exo.csv'

In [44]:
import pandas as pd
exo = pd.read_csv('data/exo.csv')
exo.count()

1. How many rows of data do you have?

In [ ]:
2. What does each row represent?

In [ ]:
3. How many planets are confirmed?

In [ ]:
4. What is the maximum temperature of the planets?

5. What is the maximum temperature of the stars?

In [ ]:
6. What is the distance between the planet and the star?

In [ ]:
7. How many planets are in each system?

In [ ]:
8. How much does the planet temperatures vary?

In [ ]:
9. How much does the star temperatures vary?

In [ ]:
10. 